website_sale_one_step_checkout/controllers/main.py

```diff
@@ -0,0 +1,240 @@
+# -*- coding: utf-8 -*-
+# © 2015 bloopark systems (<http://bloopark.de>)
+# © 2016 Antiun Ingeniería S.L. - Jairo Llopis
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+from odoo.addons.website_sale_one_step_checkout_delivery.controllers.main import WebsiteSaleOneStepCheckoutDelivery
+from odoo.http import request, redirect_with_hash
+from odoo import http, _
+from werkzeug.exceptions import Forbidden
+
+
+class WebsiteSale(WebsiteSale):
+    @http.route(['/shop/checkout'], type='http', auth='public', website=True, multilang=True)
+    def checkout(self, **post):
+        """Use one step checkout if enabled. Fall back to normal otherwise."""
+        # if onestepcheckout is deactivated use the normal checkout
+        if not request.website.use_osc:
+            return super(WebsiteSale, self).checkout(**post)
+
+        # must have a draft sale order with lines at this point, otherwise reset
+        order = request.website.sale_get_order()
+
+        redirection = self.checkout_redirection(order)
+        if redirection:
+            return redirection
+
+        values = self.checkout_values(**post)
+
+        partner = request.env['res.users'].sudo().browse(request.uid).partner_id
+
+        # get countries dependent on website settings
+        countries_domain = []
+        if not request.website.use_all_checkout_countries:
+            countries_domain = [('id', 'in', request.website.checkout_country_ids.ids)]
+
+        values['countries'] = request.env['res.country'].search(countries_domain)
+
+        # To avoid access problems when rendering
+        # the address template for public user
+        if post.get('public_user'):
+            return values
+
+        result = self.payment(post=post)
+        values.update(result.qcontext)
+
+        # Avoid useless rendering if called in ajax
```

Comment:
Do you mean the 'xhr' part? If so, yes, it is for changing the shipping address.